CRM/MembershipExtras/Form/MembershipPeriod/Base.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+use CRM_MembershipExtras_Service_ContributionUtilities as ContributionUtilities;
+
+/**
+ * Base for forms to update membership periods.
+ */
+abstract class CRM_MembershipExtras_Form_MembershipPeriod_Base extends CRM_Core_Form {
+
+  /**
+   * ID of the period to be activated.
+   *
+   * @var int
+   */
+  protected $id;
+
+  /**
+   * Maps status ID's to status names.
+   *
+   * @var array
+   */
+  protected $contributionStatusesValueMap;
+
+  /**
+   * @inheritdoc
+   */
+  public function preProcess() {
+    $this->id = CRM_Utils_Request::retrieve('id', 'String', $this, TRUE);
+    $this->contributionStatusesValueMap = ContributionUtilities::getStatusesValueMap();
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function buildQuickForm() {
+    $this->setFormTitle();
+
+    $this->addButtons([
+      [
+        'type' => 'submit',
+        'name' => E::ts('Activate'),
+        'isDefault' => TRUE,
+      ],
+      [
+        'type' => 'cancel',
+        'name' => ts('Cancel'),
+      ],
+    ]);
+
+    $period = $this->getMembershipPeriod();
+    $this->assign('period', $period);
+    $this->assign('isPaymentStarted', $this->isPaymentStarted($period));
+  }
+
+  /**
+   * Sets the title for the form.
+   *
+   * @return mixed
+   */
+  protected abstract function setFormTitle();
+
+  /**
+   * Obtains Membership Period BAO for the current period.
+   *
+   * @return \CRM_MembershipExtras_BAO_MembershipPeriod
+   */
+  protected function getMembershipPeriod() {
+    $period = new CRM_MembershipExtras_BAO_MembershipPeriod();
+    $period->id = $this->id;
+    $period->find(TRUE);
+
+    return $period;
+  }
+
+  /**
+   * Obtains payment entity status.
+   *
+   * @param \CRM_MembershipExtras_BAO_MembershipPeriod $period
+   *
+   * @return string
+   */
+  private function isPaymentStarted(CRM_MembershipExtras_BAO_MembershipPeriod $period) {
```

Comment:
have u tested this with periods without payment and that it return false correctly ?

Reply:
Tested. Works ok.